backend/de.metas.manufacturing.rest-api/src/main/java/de/metas/manufacturing/workflows_api/activity_handlers/printReceivedHUQRCodes/PrintReceivedHUQRCodesActivityHandler.java

```diff
@@ -0,0 +1,261 @@
+	private static final AdMessageKey MSG_DoYouWantToPrint = AdMessageKey.of("manufacturing.printReceivedHUQRCodes.DoYouWantToPrint?");
+	private static final AdMessageKey MSG_NoFinishedGoodsAvailableForLabeling = AdMessageKey.of("manufacturing.printReceivedHUQRCodes.NoFinishedGoodsAvailableForLabeling");
+
+	private final ManufacturingJobService manufacturingJobService;
+	private final HULabelService huLabelService;
+	private final IHUPPOrderQtyBL huPPOrderQtyBL = Services.get(IHUPPOrderQtyBL.class);
+	private final IHandlingUnitsBL handlingUnitsBL = Services.get(IHandlingUnitsBL.class);
+
+	public PrintReceivedHUQRCodesActivityHandler(
+			@NonNull final ManufacturingJobService manufacturingJobService,
+			@NonNull final HULabelService huLabelService)
+	{
+		this.manufacturingJobService = manufacturingJobService;
+		this.huLabelService = huLabelService;
+	}
+
+	@Override
+	public WFActivityType getHandledActivityType() {return HANDLED_ACTIVITY_TYPE;}
+
+	@Override
+	public UIComponent getUIComponent(final @NonNull WFProcess wfProcess, final @NonNull WFActivity wfActivity, final @NonNull JsonOpts jsonOpts)
+	{
+		return UserConfirmationSupportUtil.createUIComponent(
+				UserConfirmationSupportUtil.UIComponentProps.builderFrom(wfActivity)
+						.question(TranslatableStrings.adMessage(MSG_DoYouWantToPrint).translate(jsonOpts.getAdLanguage()))
+						.alwaysAvailableToUser(WFActivityAlwaysAvailableToUser.YES)
+						.build());
+	}
+
+	@Override
+	public WFActivityStatus computeActivityState(final WFProcess wfProcess, final WFActivity wfActivity)
+	{
+		return wfActivity.getStatus();
+	}
+
+	@Override
+	public WFProcess userConfirmed(final UserConfirmationRequest request)
+	{
+		final WFProcess wfProcess = request.getWfProcess();
+		final WFActivity wfActivity = request.getWfActivity();
+		wfActivity.getWfActivityType().assertExpected(HANDLED_ACTIVITY_TYPE);
+
+		final ImmutableList<HUToReport> hus = getHusToPrint(getPPOrderId(wfProcess));
+		printReceivedFinishedGoods(hus);
+
+		return withActivityCompleted(wfProcess, wfActivity);
+	}
+
+	private WFProcess withActivityCompleted(final WFProcess wfProcess, final WFActivity wfActivity)
+	{
+		final ManufacturingJob job = getJob(wfProcess);
+		final ManufacturingJobActivityId jobActivityId = wfActivity.getId().getAsId(ManufacturingJobActivityId.class);
+		final ManufacturingJob changedJob = manufacturingJobService.withActivityCompleted(job, jobActivityId);
+		return ManufacturingRestService.toWFProcess(changedJob);
+	}
+
+	@NonNull
+	private static ManufacturingJob getJob(final WFProcess wfProcess)
+	{
+		return wfProcess.getDocumentAs(ManufacturingJob.class);
+	}
+
+	private PPOrderId getPPOrderId(@NonNull final WFProcess wfProcess)
+	{
+		return getJob(wfProcess).getPpOrderId();
+	}
+
+	private ImmutableList<HUToReport> getHusToPrint(final PPOrderId ppOrderId)
+	{
+		final Set<HuId> huIds = huPPOrderQtyBL.getFinishedGoodsReceivedHUIds(ppOrderId);
```

Comment:
IMHO we shall have a method ManufacturingJobService.getFinishedGoodsReceivedHUIds(ppOrderId) which calls huPPOrderQtyBL 

... we can keep that abstracted.